### PR TITLE
feat() : #Eval 464  add notes in csv

### PR DESCRIPTION
**Pull request by lapeyronnie**
## Describe your changes
ajout des notes des élèves pour chaque devoir de la période et de la matière
refacto : magic string de la competence -> Field
## Checklist tests
export csv avec des devoirs sur la période, avec des notes ou des annotations
export csv sans devoir
## Issue ticket number and link
[EVAL-464](https://jira.support-ent.fr/browse/EVAL-464) première partie du ticket [EVAL-294](https://jira.support-ent.fr/browse/EVAL-294)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [ ] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java [hunk 1]
```diff
@@ -2019,27 +2019,29 @@ private void setAverageOfSubjects(JsonObject result, List<Service> services, Jso
...
-                matiereJO.put("moyenne", (moyenneFinale.getString("moyenne") != null)? moyenneFinale.getString("moyenne"): NN);
+                matiereJO.put("moyenne", (moyenneFinale.getString("moyenne") != null)? moyenneFinale.getString("moyenne"): Field.NN);
             } else if (!notes.isEmpty() || subTopicNoteDevoirMap.size() > 0) {
                 if(subTopicNoteDevoirMap.size()> 0){
                     AtomicReference<Double> coefTotal = new AtomicReference<>(0.d);
                     AtomicReference<Double> total = new AtomicReference<>(0.d);
                     subTopicNoteDevoirMap.forEach((subtopic,notesList) ->{
-                        total.updateAndGet(v -> v + utilsService.calculMoyenne(notesList, false, 20, false).getDouble("moyenne") * subtopic.getCoefficient());
+                        total.updateAndGet(v -> v + utilsService.calculMoyenne(notesList, false, 20,
+                                false).getDouble("moyenne") * subtopic.getCoefficient());
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java:2029]: Magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java [hunk 2]
```diff
@@ -2019,27 +2019,29 @@ private void setAverageOfSubjects(JsonObject result, List<Service> services, Jso
...
-                        matiereJO.put("moyenne", NN);
+                        matiereJO.put("moyenne", Field.NN);
                     }else{
                         matiereJO.put("moyenne", decimalFormat.format(total.get() / coefTotal.get()));
                     }
                 }else {
 
-                    Double moy = utilsService.calculMoyenne(notes, false, 20, false).getDouble("moyenne");
+                    Double moy = utilsService.calculMoyenne(notes, false, 20, false)
+                            .getDouble("moyenne");
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java:2040]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java [hunk 3]
```diff
@@ -2019,27 +2019,29 @@ private void setAverageOfSubjects(JsonObject result, List<Service> services, Jso
...
                     matiereJO.put("moyenne", decimalFormat.format(moy));
                 }
             } else {
-                matiereJO.put("moyenne", NN);
+                matiereJO.put("moyenne", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultDevoirService.java:2044]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 4]
```diff
@@ -3415,7 +3412,7 @@ private <T> void calculMoyennesNotesForReleve(JsonArray listNotes, JsonObject re
                 result.put("moyenne_classe", moyClasse);
             }
             result.put("_moyenne_classe", new JsonObject());
-            JsonObject moyClasseObj = new JsonObject().put("min", min).put("max", max).put(MOYENNE, moyClasse);
+            JsonObject moyClasseObj = new JsonObject().put("min", min).put("max", max).put(Field.MOYENNE, moyClasse);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3415]: Magic String

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 5]
```diff
@@ -3424,7 +3421,7 @@ private <T> void calculMoyennesNotesForReleve(JsonArray listNotes, JsonObject re
                 int nbSousMoyClass = mapNbMoyenneClasse.get(idSousMat);
                 Object moySous = (nbSousMoyClass > 0) ? (moySousMat / nbSousMoyClass) : " ";
                 JsonObject moySousMatCl = new JsonObject().put("min", mapMin.get(idSousMat))
-                        .put("max", mapMax.get(idSousMat)).put(MOYENNE, decimalFormat.format(moySous));
+                        .put("max", mapMax.get(idSousMat)).put(Field.MOYENNE, decimalFormat.format(moySous));
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3424]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 6]
```diff
@@ -3532,36 +3529,36 @@ private <T> void calculMoyennesNotesForReleve(JsonArray listNotes, JsonObject re
...
                                         if (moyObj != null) {
-                                            moyenne.remove(MOYENNE);
-                                            if ((moyObj.getString(MOYENNE) != null)) {
-                                                moyenne.put(MOYENNE, moyObj.getString(MOYENNE));
+                                            moyenne.remove(Field.MOYENNE);
+                                            if ((moyObj.getString(Field.MOYENNE) != null)) {
+                                                moyenne.put(Field.MOYENNE, moyObj.getString(Field.MOYENNE));
                                             } else {
-                                                moyenne.put(MOYENNE, "NN");
+                                                moyenne.put(Field.MOYENNE, "NN");
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3554]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 7]
```diff
@@ -3796,10 +3793,10 @@ private void getMoyenneMinMaxByMatiere(Map<String, JsonObject> eleveMapObject, L
 
         JsonObject statsToAdd = new JsonObject();
         if(!initialisationPositionnement){
-            JsonObject minPos = new JsonObject().put("minimum", NN);
+            JsonObject minPos = new JsonObject().put("minimum", Field.NN);
             statsToAdd.put("positionnement",minPos);
-            statsToAdd.getJsonObject("positionnement").put("maximum", NN);
-            statsToAdd.getJsonObject("positionnement").put("moyenne", NN);
+            statsToAdd.getJsonObject("positionnement").put("maximum", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3798]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 8]
```diff
@@ -3796,10 +3793,10 @@ private void getMoyenneMinMaxByMatiere(Map<String, JsonObject> eleveMapObject, L
...
+            statsToAdd.getJsonObject("positionnement").put("moyenne", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3799]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 9]
```diff
@@ -3815,10 +3812,10 @@ private void getMoyenneMinMaxByMatiere(Map<String, JsonObject> eleveMapObject, L
 
         }
         if(!initialisationMoyenne){
-            JsonObject minPos = new JsonObject().put("minimum", NN);
+            JsonObject minPos = new JsonObject().put("minimum", Field.NN);
             statsToAdd.put("moyenne",minPos);
-            statsToAdd.getJsonObject("moyenne").put("maximum", NN);
-            statsToAdd.getJsonObject("moyenne").put("moyenne", NN);
+            statsToAdd.getJsonObject("moyenne").put("maximum", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3817]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 10]
```diff
@@ -3965,7 +3962,7 @@ private void getMoyenneGeneraleMinMax(Map<String, JsonObject> eleveMapObject, Lo
             }
 
             if(nbMatieres == 0){
-                student.getValue().put("moyenne_generale", NN);
+                student.getValue().put("moyenne_generale", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3965]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 11]
```diff
@@ -3989,10 +3986,10 @@ private void getMoyenneGeneraleMinMax(Map<String, JsonObject> eleveMapObject, Lo
         }
 
         if(nbElevesMoyenne == 0){
-            JsonObject minMoy = new JsonObject().put("minimum", NN);
+            JsonObject minMoy = new JsonObject().put("minimum", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3989]: magic string

File: src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java [hunk 12]
```diff
@@ -3989,10 +3986,10 @@ private void getMoyenneGeneraleMinMax(Map<String, JsonObject> eleveMapObject, Lo
...
             resulHandler.getJsonObject("statistiques").put("moyenne_generale",minMoy);
-            resulHandler.getJsonObject("statistiques").getJsonObject("moyenne_generale").put("maximum", NN);
-            resulHandler.getJsonObject("statistiques").getJsonObject("moyenne_generale").put("moyenne", NN);
+            resulHandler.getJsonObject("statistiques").getJsonObject("moyenne_generale").put("maximum", Field.NN);
```
Comment (Rseuret) [inline src/main/java/fr/openent/competences/service/impl/DefaultNoteService.java:3991]: magic string

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts [hunk 13]
```diff
@@ -429,6 +428,12 @@ export class ReleveNote extends  Model implements IModel {
         });
     }
 
+    private getAnnotation(idAnnotation : number) : string {
+        var lib = evaluations.structure.annotations.findWhere(
```
Comment (Rseuret) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:432]: typage + let

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts [hunk 14]
```diff
@@ -583,6 +566,17 @@ export class ReleveNote extends  Model implements IModel {
         return line[key]
     }
 
+    private setNoteOnDevoir (line : any, notes : any) {
```
Comment (Rseuret) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:569]: pas moyen de typer avec autre chose que "any"?

**Review by Rseuret — CHANGES_REQUESTED** [review]

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts (hunk 14 above)
Comment (Rseuret) [reply src/main/resources/public/ts/models/teacher/ReleveNote.ts:569]: typage fonction

Comment (Rseuret) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:569]:
```suggestion
    private setNoteOnDevoir (line : any, notes : any) : void {
```

**Comment by Rseuret** [comment]
feat() : #Eval 464 add notes in csv ->  feat() #Eval-464 : add notes in csv

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts (hunk 14 above)
Comment (lapeyronnie) [reply src/main/resources/public/ts/models/teacher/ReleveNote.ts:569]: oui pour notes mais pas pour line

**Review by Rseuret — APPROVED** [review]

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts [hunk 15]
```diff
@@ -713,11 +716,24 @@ export class ReleveNote extends  Model implements IModel {
             };
         });
     }
+    IsSpecificOptionsCsv () : boolean {
+        return this.exportOptions.withNotes;
+    }
+
+    setTypeExport ()  {
```
Comment (JinDouang) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:723]:
```suggestion
    setTypeExport(): void  {
```

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts [hunk 16]
```diff
@@ -713,11 +716,24 @@ export class ReleveNote extends  Model implements IModel {
...
+        if (this.IsSpecificOptionsCsv) this.exportOptions.fileType = 'csv';
+    }
 
     formateHeaderAndColumn () : any {
         let header = `${lang.translate('students')}`;
         let column = ['displayName'];
-
+        if(this.exportOptions.withNotes) {
+            _.each(this.devoirs.all, (d) => {
+                let title = d.name;
```
Comment (JinDouang) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:732]: typage

```suggestion
                let title: string = d.name;
```

File: src/main/resources/public/ts/models/teacher/ReleveNote.ts [hunk 17]
```diff
@@ -726,6 +742,10 @@ export class ReleveNote extends  Model implements IModel {
             header +=`; ${lang.translate('average.final')}`;
             column.push('moyenneFinale');
         }
+        if(this.exportOptions.withNotes) {
+            header +=`; ${lang.translate('average')}`;
+            column.push('actualAverage')
```
Comment (JinDouang) [inline src/main/resources/public/ts/models/teacher/ReleveNote.ts:747]:
```suggestion
            column.push('actualAverage');
```

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**